File: algorithms/442. Find All Duplicates in an Array/solution.py
class Solution(object):
    def findDuplicates(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        dup = []
        for n in nums:
            idx = abs(n) - 1
            if nums[idx] < 0:
                dup.append(idx + 1)
            else:
                nums[idx] = -nums[idx]
        return dup

# 27 / 27 test cases passed.
# Status: Accepted
# Runtime: 308 ms
# 82.47%

File: algorithms/442. Find All Duplicates in an Array/solution.cc
class Solution {
public:
    vector<int> findDuplicates(vector<int>& nums) {
        vector<int> dup;
        for ( int i = 0; i < nums.size(); ++i ) {
            if ( nums[abs(nums[i]) - 1] < 0 ) {
                dup.push_back(abs(nums[i]));
            }
            nums[abs(nums[i]) - 1] *= -1;
        }
        return dup;
    }
};
